smartpy/losses/classification_losses.py

```diff
@@ -0,0 +1,16 @@
+from smartpy.interfaces.loss import Loss
+
+import theano.tensor as T
+
+
+class NegativeLogLikelihood(Loss):
+    def _loss_function(self, model_output):
+        nll = -T.log(model_output)
+        indices = T.cast(self.target[:, 0], dtype="int32")  # Targets are floats.
+        selected_nll = nll[T.arange(self.target.shape[0]), indices]
```

Comment:
I mean for building the matrix. Passing an array as argument would require to iterate over that array. Slicing with one of the two snippet of code I've written would require only constant time.

Consider this : 

    In [1]: import numpy as np
    
    In [2]: A = np.random.random((10000000,1))
    
    In [3]: indices = np.arange(10000000)
    
    In [4]: %time A[indices,0]
    CPU times: user 20 ms, sys: 0 ns, total: 20 ms
    Wall time: 19.7 ms
    Out[4]: 
    array([ 0.52468624,  0.38960097,  0.22404771, ...,  0.55581394,
            0.45208798,  0.49348217])
    
    In [5]: %time A.T[0]
    CPU times: user 0 ns, sys: 0 ns, total: 0 ns
    Wall time: 8.34 µs
    Out[5]: 
    array([ 0.52468624,  0.38960097,  0.22404771, ...,  0.55581394,
            0.45208798,  0.49348217])
    
    In [6]: %time A[:,0].reshape((A.shape[0],))
    CPU times: user 0 ns, sys: 0 ns, total: 0 ns
    Wall time: 10 µs
    Out[6]: 
    array([ 0.52468624,  0.38960097,  0.22404771, ...,  0.55581394,
            0.45208798,  0.49348217])

Your method is about 2000 times slower, probably because it's *O(n)*.